File: ask-o11y/pkg/rbac/rbac.go
package rbac

import "consensys-asko11y-app/pkg/mcp"

func IsReadOnlyTool(tool mcp.Tool) bool {
	if tool.Annotations == nil || tool.Annotations.ReadOnlyHint == nil {
		return false
	}
	return *tool.Annotations.ReadOnlyHint
}

func CanAccessTool(role string, tool mcp.Tool) bool {
	if role == "Admin" || role == "Editor" {
		return true
	}
	return IsReadOnlyTool(tool)
}

func FilterToolsByRole(tools []mcp.Tool, role string) []mcp.Tool {
	if role == "Admin" || role == "Editor" {
		return tools
	}
	filtered := make([]mcp.Tool, 0, len(tools))
	for _, t := range tools {
		if CanAccessTool(role, t) {
			filtered = append(filtered, t)
		}
	}
	return filtered
}

File: ask-o11y/src/utils/rbac.ts
/**
 * RBAC utilities for filtering MCP tools by user role using tool annotations.
 *
 * Role hierarchy:
 * - Admin/Editor: Full access to all tools
 * - Viewer: Read-only access (tools with readOnlyHint annotation)
 */

export type UserRole = 'Admin' | 'Editor' | 'Viewer';

interface ToolWithAnnotations {
  name: string;
  annotations?: {
    readOnlyHint?: boolean;
  };
}

export function isReadOnlyTool(tool: ToolWithAnnotations): boolean {
  return tool.annotations?.readOnlyHint === true;
}

export function canAccessTool(role: UserRole | string, tool: ToolWithAnnotations): boolean {
  if (role === 'Admin' || role === 'Editor') {
    return true;
  }
  return isReadOnlyTool(tool);
}

export function filterToolsByRole<T extends ToolWithAnnotations>(tools: T[], role: UserRole | string): T[] {
  if (role === 'Admin' || role === 'Editor') {
    return tools;
  }
  return tools.filter((tool) => canAccessTool(role, tool));
}
